src/ILCompiler.MetadataTransform/src/ILCompiler/Metadata/ExplicitScopeAssemblyPolicyMixin.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Cts = Internal.TypeSystem;
+using Ecma = System.Reflection.Metadata;
+
+namespace ILCompiler.Metadata
+{
+    /// <summary>
+    /// Mixin helper class to allow easy support for ExplicitScopeAttribute for metadata transformation
+    /// 
+    /// ExplicitScopeAttribute is used to relocate where a given type appears in metadata from one 
+    /// metadata assembly to another. It must not be used to relocate into an assembly which otherwise 
+    /// exists within the compilation operation (Current implementation is not reliable in those circumstances.)
+    /// </summary>
+    public class ExplicitScopeAssemblyPolicyMixin
+    {
+        private class AssemblyNameEqualityComparer : IEqualityComparer<AssemblyName>
+        {
+            public static AssemblyNameEqualityComparer Instance { get; } = new AssemblyNameEqualityComparer();
+
+            public bool ByteArrayCompare(byte[] arr1, byte[] arr2)
+            {
+                if (arr1.Length != arr2.Length)
+                    return false;
+
+                for (int i = 0; i < arr1.Length; i++)
+                {
+                    if (arr1[i] != arr2[i])
+                        return false;
+                }
+                return true;
+            }
+
+            public bool Equals(AssemblyName x, AssemblyName y)
+            {
+                if (x.Name != y.Name)
+                    return false;
+
+                if (x.ContentType != y.ContentType)
+                    return false;
+
+                if (x.CultureName != y.CultureName)
+                    return false;
+
+                if (x.Flags != y.Flags)
+                    return false;
+                
+                if (x.Flags.HasFlag(AssemblyNameFlags.PublicKey))
+                {
+                    if (!ByteArrayCompare(x.GetPublicKey(), y.GetPublicKey()))
+                        return false;
+                }
+                else
+                {
+                    if (!ByteArrayCompare(x.GetPublicKeyToken(), y.GetPublicKeyToken()))
+                        return false;
+                }
+
+                if (x.ProcessorArchitecture != y.ProcessorArchitecture)
+                    return false;
+
+                if (!x.Version.Equals(y.Version))
+                    return false;
+
+                return true;
+            }
+
+            public int GetHashCode(AssemblyName obj)
+            {
+                return obj.Name.GetHashCode();
+            }
+        }
+
+        private class ExplicitScopeAssemblyDesc : Cts.ModuleDesc, Cts.IAssemblyDesc
```

Comment:
Nit: We only add the `Desc` suffix to the base type (so, Module*Desc* and Type*Desc*, but Ecma*Module* and Ecma*Type*)